crates/shielder-account/src/secrets.rs

```diff
@@ -28,22 +39,34 @@ fn hash(id: U256, label: &[u8], payload: &[u8]) -> U256 {
 pub mod nonced {
     use alloy_primitives::U256;
 
-    use super::{hash, NULLIFIER_LABEL, TRAPDOOR_LABEL};
+    use super::{hash, Label};
 
     /// Returns a pseudorandom field element deterministically computed from `id` and `nonce`.
     pub fn derive_nullifier(id: U256, nonce: u32) -> U256 {
-        hash(id, NULLIFIER_LABEL, &nonce.to_be_bytes())
+        let mut payload = Vec::new();
+        payload.extend(id.to_be_bytes_vec());
+        payload.extend_from_slice(&nonce.to_be_bytes());
+        hash(Label::Nullifier, &payload)
     }
 
     /// Returns a pseudorandom field element deterministically computed from `id` and `nonce`.
     pub fn derive_trapdoor(id: U256, nonce: u32) -> U256 {
-        hash(id, TRAPDOOR_LABEL, &nonce.to_be_bytes())
+        let mut payload = Vec::new();
+        payload.extend(id.to_be_bytes_vec());
+        payload.extend_from_slice(&nonce.to_be_bytes());
+        hash(Label::Trapdoor, &payload)
     }
 }
 
 /// Private-key-dependent derivation of a per-token private ID.
-pub fn derive_id(private_key: U256, token_address: Address) -> U256 {
-    hash(private_key, ID_LABEL, token_address.into_word().as_ref())
+pub fn derive_id(private_key: U256, chain_id: u64, token_address: Address) -> U256 {
+    // Concatenate chain_id bytes and token_address bytes
+    let mut payload = Vec::new();
+    payload.extend(private_key.to_be_bytes_vec());
+    payload.extend_from_slice(&chain_id.to_be_bytes());
+    payload.extend_from_slice(token_address.into_word().as_ref());
```

Comment:
according to the docs, we should get here account nonce instead of token address, shouldn't we?